pynumdiff/optimize/_optimize.py

```diff
@@ -0,0 +1,186 @@
+                    is not None, see _objective_function
+    :param str opt_method: Optimization technique used by :code:`scipy.minimize`, the workhorse
+    :param int maxiter: passed down to :code:`scipy.minimize`, maximum iterations
+
+    :return: tuple[dict, float] of\n
+            - **opt_params** -- best parameter settings for the differentation method
+            - **opt_value** -- lowest value found for objective function
+    """
+    if metric not in ['rmse','error_correlation']:
+        raise ValueError('`metric` should either be `rmse` or `error_correlation`.')
+    if metric == 'error_correlation' and dxdt_truth is None:
+        raise ValueError('`metric` can only be `error_correlation` if `dxdt_truth` is given.')
+
+    params, bounds = method_params_and_bounds[func]
+    params.update(search_space) # for things not given, use defaults
+
+    # No need to optimize over singletons, just pass them through
+    singleton_params = {k:v for k,v in params.items() if not isinstance(v, list)}
+
+    # The search space is the Cartesian product of all dimensions where multiple options are given
+    search_space_types = {k:type(v[0]) for k,v in params.items() if isinstance(v, list)} # for converting back and forth from point
+    if any(v not in [float, int, bool] for v in search_space_types.values()):
+        raise ValueError("Optimization over categorical strings currently not supported")
+    # If excluding string type, I can just cast ints and bools to floats, and we're good to go
+    cartesian_product = product(*[np.array(params[k]).astype(float) for k in search_space_types])
+    
+    bounds = [bounds[k] if k in bounds else # pass these to minimize(). It should respect them.
+             (0, 1) if v == bool else
+             None for k,v in search_space_types.items()]
+
+    # wrap the objective and scipy.optimize.minimize because the objective and options are always the same
+    _obj_fun = partial(_objective_function, func=func, x=x, dt=dt, singleton_params=singleton_params,
+        search_space_types=search_space_types, dxdt_truth=dxdt_truth, metric=metric, tvgamma=tvgamma,
+        padding=padding)
+    _minimize = partial(scipy.optimize.minimize, _obj_fun, method=opt_method, bounds=bounds, options={'maxiter':maxiter})
+
+    with Pool(initializer=filterwarnings, initargs=["ignore", '', UserWarning]) as pool: # The heavy lifting
+        results = pool.map(_minimize, cartesian_product) # returns a bunch of OptimizeResult objects
+
+    opt_idx = np.nanargmin([r.fun for r in results])
+    opt_point = results[opt_idx].x
+    # results are going to be floats, but that may not be allowed, so convert back to a dict
+    opt_params = {k:(v if search_space_types[k] == float else 
+                    int(np.round(v)) if search_space_types[k] == int else
+                    v > 0.5) for k,v in zip(search_space_types, opt_point)}
```

Comment:
This translation-back-to-dict code is awfully compressed. It may need to become a full-on encode/decode function pair if we extend this too handle strings. But I'm not sure we're going to do that, or if we do, we may take a different approach, like Bayesian optimization with a library that handles this sort of thing a bit.